### PR TITLE
lease,systemschema: hook up the lease table to regional-by-row partitioning

### PR DESCRIPTION
**Pull request by ajwerner**
This PR extends the `system.lease` table with REGIONAL BY ROW partitioning for use with https://github.com/cockroachdb/cockroach/pull/92395. The PR is itself freestanding. 

Epic: CRDB-18596

Release note: None

### CODE REVIEW

File: pkg/server/server_sql.go [hunk 1]
```diff
@@ -1298,6 +1298,8 @@ func (s *SQLServer) preStart(
 		return err
 	}
 
+	s.leaseMgr.SetRegionPrefix(regionPhysicalRep)
```
Comment (JeffSwenson) [inline pkg/server/server_sql.go:1301]: regionPhysicalRep may be []byte{} and if it is, we should use enum.One instead. We discussed pushing this into GetLocalityRegionEnumPhysicalRepresentation, but I think that would be surprising behavior for the API. Instead, we could put the branch in server_sql. Alternatively we could defer initializing the leasing subsystem until a sql session is created and the lease could use the region from the session.

If we don't use enum.One, then we run into problems using sql to query leases that were created before the system database is configured as multi-region:

```
jeff@multiregion-8.6s4p.crdb.io:26257/defaultdb> select * from system.lease;
ERROR: internal error: unexpected error from the vectorized engine: could not find [] in enum "system.public.crdb_internal_region" representation PhysicalReps: [[64] [96] [128]]; LogicalReps: [gcp-asia-southeast1 gcp-europe-west1 gcp-us-central1] goroutine 12874 [running]:
```

File: pkg/sql/catalog/systemschema/system.go [hunk 2]
```diff
@@ -1215,29 +1225,65 @@ var (
...
+						{
+							Name:        "primary",
+							ID:          0,
+							ColumnNames: []string{"descID", "version", "nodeID", "expiration", "crdb_region"},
+							ColumnIDs:   []descpb.ColumnID{1, 2, 3, 4, 5},
+						},
+					},
+					descpb.IndexDescriptor{
+						Name:           "primary",
+						ID:             1,
```
Comment (JeffSwenson) [inline pkg/sql/catalog/systemschema/system.go:1252]: The multi-region index should use ID 2 if we want to support migrating from the old schema to the new schema using a protocol resembling a standard schema change.

File: pkg/server/server_sql.go (hunk 1 above)
Comment (ajwerner) [reply pkg/server/server_sql.go:1301]: This is handled inside that call. It has the below logic. The value is initialized to enum.One.
```go
if len(val) > 0 {
		m.storage.regionPrefix.Store(val)
}
```

File: pkg/sql/catalog/lease/descriptor_state.go [hunk 3]
```diff
@@ -130,7 +130,7 @@ func (t *descriptorState) findForTimestamp(
 // If an existing lease exists for the descriptor version it replaces
 // it and returns it.
 func (t *descriptorState) upsertLeaseLocked(
-	ctx context.Context, desc catalog.Descriptor, expiration hlc.Timestamp,
+	ctx context.Context, desc catalog.Descriptor, expiration hlc.Timestamp, prefix []byte,
```
[the remark this answers is not in the document]
Comment (ajwerner) [reply pkg/sql/catalog/lease/descriptor_state.go:133]: Done.

File: pkg/sql/catalog/lease/lease.go [hunk 4]
```diff
@@ -730,6 +732,8 @@ func NewLeaseManager(
 		stopper:          stopper,
 		sem:              quotapool.NewIntPool("lease manager", leaseConcurrencyLimit),
 	}
+	lm.storage.regionPrefix = &atomic.Value{}
+	lm.storage.regionPrefix.Store(enum.One)
```
[the remark this answers is not in the document]
Comment (ajwerner) [reply pkg/sql/catalog/lease/lease.go:736]: We prevent it from being the empty slice. We want it to always be a value. Enums always have this value unless it has been removed.

File: pkg/sql/catalog/lease/lease.go [hunk 5]
```diff
@@ -759,6 +763,22 @@ func (m *Manager) findNewest(id descpb.ID) *descriptorVersionState {
 	return t.mu.active.findNewest()
 }
 
+// SetRegionPrefix sets the prefix this Manager uses to write leases. If val
+// is empty, this call is a no-op. Note that the default value is enum.One.
+// This means that leases acquired before initial startup may write their
```
[the remark this answers is not in the document]
Comment (ajwerner) [reply pkg/sql/catalog/lease/lease.go:768]: In an non-MR cluster this doesn't do anything.

File: pkg/sql/catalog/systemschema/system.go (hunk 2 above)
Comment (ajwerner) [reply pkg/sql/catalog/systemschema/system.go:1252]: Done.

**Review by chengxiong-ruan — APPROVED** [review]

**Comment by ajwerner** [comment]
TFTR!

bors r+

**Comment by ajwerner** [comment]
bors r+ 

flake was https://github.com/cockroachdb/cockroach/pull/93287

**Pull request merged**